### ISSUE TITLE
Requested operation has changed during the UI authentication session - when trying to deactivate user with erasure

### ISSUE THREAD

**Issue by Twi1ightSparkle** [matrix-org/synapse#7471]
### Description

Deactivating an account with "Please forget all messages I have sent when my account is deactivated" checked does not work. You get the error "There was a problem communicating with the server. Please try again."

Network tab show 401 for `deactivate`, then 200, then 403.

The spinner thing goes on forever (at least a couple of hours before I gave up)

I tested this with an almost new account with very few messages. A customer also reported this, awaiting rageshake from them. My rageshake is linked

https://newvector.zammad.com/#ticket/zoom/3216
(customer email copy-pasted to my rageshake)

### Steps to reproduce

- Create a matrix.org
- Settings
- Deactivate Account
- Check "Please forget all messages I have sent when my account is deactivated"
- Enter password
- Click Continue

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

Matrix.org, account deletion does work on my Modular server

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**: matrix.org

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**: "1.12.4 (b=matrix-org-hotfixes,309e30bae)"

- **Install method**:
<!-- examples: package manager/git clone/pip  -->

- **Platform**:
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

**Comment by Twi1ightSparkle** [comment]
Synapse logs added to rageshake

**Comment by anoadragon453** [comment]
So this is happening because Riot starts the User-Interactive Authentication Session with:

```json
{"erase": false}
```

then, after the erase checkbox is checked and the user enters their password, 

```json
{
  "auth": {
    "session": "yyy",
    "type": "m.login.password",
    "user": "@xxx:matrix.org",
    "identifier": {
      "type": "m.id.user",
      "user": "@xxx:matrix.org"
    },
    "password": "xxx"
  },
  "erase": true
}
```

Now that `erase` is `true`, Synapse complains with:

```json
{
  "errcode": "M_UNKNOWN",
  "error": "Requested operation has changed during the UI authentication session."
}
```

I'm not sure if this should be initially fixed from the Synapse or Riot side, but we'll need to inform clients that there's a potential breaking change if they were relying on changing the UIAA parameters mid-way through an authentication session, as Riot is doing here.

This check was introduced in https://github.com/matrix-org/synapse/pull/7068. Unfortunately, even after https://github.com/matrix-org/synapse/pull/7455 which relaxed the requirements a bit, this is still a problem in certain cases (but again, maybe a client problem rather than a server one).

This occurs on matrix.org as it is running v1.13.0rc - the check is not a part of v1.12.4.

**Comment by richvdh** [comment]
I feel pretty strongly that riot is at fault for (a) moving the goalposts mid-operation and (b) not giving the user a better error

**Comment by clokep** [comment]
See also #7452.

**Comment by clokep** [comment]
This should be fixed in 1.13.0 as of #7483.
